### PR TITLE
Add documentation - production checklist / Deprecate ClientFactoryBuilder.socketOption() over channelOption()

### PR DESCRIPTION
**Pull request by trustin**
Motivation:

A simple checklist before putting an Armeria application into production
would be very useful for our users.

Modifications:

- Added `advanced-production-checklist.rst` which is based on the recent
  Slack thread with @jwills, @anuraaga and @huydx
- Added how to use the `CircuitBreaker` API with non-Armeria clients
- Deprecate ClientFactoryBuilder.socketOption() with channelOption() for
  consistency
- Use Object2ObjectArrayMap for storing channel options

Result:

- Better UX
- A little bit of efficiency

/cc @jwills

### CODE REVIEW

File: site/src/sphinx/client-circuit-breaker.rst [hunk 1]
```diff
@@ -269,3 +269,40 @@ If you use :api:`CircuitBreakerBuilder`, you can configure the parameters which
...
+
+Using ``CircuitBreaker`` with non-Armeria client
+------------------------------------------------
+
+:api:`CircuitBreaker` API is designed to be framework-agnostic and thus can be used with any non-Armeria
+clients:
+
+1. Create a :api:`CircuitBreaker`.
+2. Guard your client calls with ``CircuitBreaker.canRequest()``.
+3. Update the state of :api:`CircuitBreaker` by calling ``CircuitBreaker.onSuccess()`` or ``onFailure``
```
Comment (minwoox) [inline site/src/sphinx/client-circuit-breaker.rst]: `onFailure()` for consistency

Comment (trustin) [reply site/src/sphinx/client-circuit-breaker.rst]: Fixed

File: site/src/sphinx/advanced-production-checklist.rst [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    Note that the advices in this page are not always applicable for every use case and thus should be
+    applied with caution. Do not apply the changes you really do not need.
+
+You may want to consider the following options before putting your Armeria application into production.
+
+- Specify the maximum number of accepted connections. The default is *unbounded*.
+
+  .. code-block:: java
+
```
[the remark this answers is not in the document]
Comment (trustin) [reply site/src/sphinx/advanced-production-checklist.rst]: Added

File: site/src/sphinx/advanced-production-checklist.rst [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      ServerBuilder sb = new ServerBuilder();
+      sb.blockingTaskExecutor(myBoundedExecutor);
+
+- Specify the default limits of an HTTP request or response.
+
+  .. code-block:: java
+
```
[the remark this answers is not in the document]
Comment (trustin) [reply site/src/sphinx/advanced-production-checklist.rst]: Added

**Comment by huydx** [comment]
nits: how about mention briefly about recently added customizable ChannelOption (both Server and Client sides)

**Comment by trustin** [comment]
@huydx Updated the PR. Please review again.

**Comment by trustin** [comment]
Thanks, reviewers!

**Pull request merged**